api/v1alpha1/stargate_types.go

```diff
@@ -65,26 +58,85 @@ type StargateTemplate struct {
 	// +optional
 	ReadinessProbe *corev1.Probe `json:"readinessProbe,omitempty"`
 
-	// Affinity is the affinity to apply to the Stargate pods. See
+	// Affinity is the affinity to apply to all the Stargate pods. See
 	// https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
-	// for background.
+	// for background. Leave nil to let the controller reuse the same affinity rules used for data
+	// nodes in this datacenter.
 	// +optional
 	Affinity *corev1.Affinity `json:"affinity,omitempty"`
 
 	// Tolerations are tolerations to apply to the Stargate pods. See
-	//# https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/ for background.
+	// https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/ for background.
+	// Leave nil to let the controller reuse the same tolerations used for data nodes in this datacenter, if any.
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
 
-	// CassandraConfigMap is a reference to a ConfigMap that holds Cassandra configuration.
+	// CassandraConfigMapRef is a reference to a ConfigMap that holds Cassandra configuration.
 	// The map should have a key named cassandra_yaml.
 	// +optional
-	CassandraConfigMap *corev1.LocalObjectReference `json:"cassandraConfigMap,omitempty"`
+	CassandraConfigMapRef *corev1.LocalObjectReference `json:"cassandraConfigMap,omitempty"`
+}
+
+// StargateClusterTemplate defines global rules to apply to all Stargate pods in all datacenters in the cluster.
+// These rules will be merged with rules defined at datacenter level in a StargateDatacenterTemplate; dc-level rules
+// have precedence over cluster-level ones.
+type StargateClusterTemplate struct {
+	StargateTemplate `json:",inline"`
+
+	// Size is the number of Stargate instances to deploy in each datacenter. They will be spread evenly across racks.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=1
+	Size int32 `json:"size"`
+}
+
+// StargateDatacenterTemplate defines rules to apply to all Stargate pods in a given datacenter.
+// These rules will be merged with rules defined at rack level in a StargateRackTemplate; rack-level rules
+// have precedence over datacenter-level ones.
+type StargateDatacenterTemplate struct {
+	StargateClusterTemplate `json:",inline"`
+
+	// Racks allow customizing Stargate characteristics for specific racks in the datacenter.
+	// +optional
+	Racks []StargateRackTemplate `json:"racks,omitempty"`
+}
+
+// MergeWith merges this StargateDatacenterTemplate with the given StargateClusterTemplate.
+func (in *StargateDatacenterTemplate) MergeWith(clusterTemplate *StargateClusterTemplate) *StargateDatacenterTemplate {
+	if in == nil && clusterTemplate == nil {
+		return nil
+	} else if in == nil {
+		return &StargateDatacenterTemplate{StargateClusterTemplate: *clusterTemplate}
+	} else {
+		return in
```

Comment:
It's more of an override than a merge here: if the dc-level template is non nil it will be used, otherwise if the cluster-level template is non nil it will be used, otherwise Stargate won't be deployed at all. 

As discussed offline, this has the merit of being simple and straightforward. But I would see a case for merging templates down; consider the following definition:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: Stargate
metadata:
  name: s1
spec:
  datacenterRef:
    name: dc1
  size: 3
  heapSize: 512Mi
  nodeSelector:
      - foo: bar
  racks:
    - name: rack1
      nodeSelector:
        - foo: qix
```

Here, rack1 has its own template with a different node selector than the DC-level one. However, I would expect rack1 to also inherit `heapSize:512Mi`, but that's not how it works currently: in rack1, the heap size would be the default one (`256Mi`).

Reply:
I am in favor of sticking with what we have for now, getting this PR merged, and then start a discussion on the forum to solicit input as to whether merge behavior would be preferred.

Reply:
Sounds good. Once this PR is merged I will compile a list of follow-up actions, which will include creating a few GH issues and hitting the public forum for input on this topic.